runtime/cni_integ_test.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/containerd/oci"
+	"github.com/containerd/containerd/pkg/ttrpcutil"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/firecracker-microvm/firecracker-containerd/internal"
+	"github.com/firecracker-microvm/firecracker-containerd/proto"
+	fccontrol "github.com/firecracker-microvm/firecracker-containerd/proto/service/fccontrol/ttrpc"
+	"github.com/firecracker-microvm/firecracker-containerd/runtime/firecrackeroci"
+)
+
+func TestCNISupport_Isolated(t *testing.T) {
+	internal.RequiresIsolation(t)
+
+	testTimeout := 120 * time.Second
+	ctx, cancel := context.WithTimeout(namespaces.WithNamespace(context.Background(), defaultNamespace), testTimeout)
+	defer cancel()
+
+	client, err := containerd.New(containerdSockPath, containerd.WithDefaultRuntime(firecrackerRuntime))
+	require.NoError(t, err, "unable to create client to containerd service at %s, is containerd running?", containerdSockPath)
+	defer client.Close()
+
+	pluginClient, err := ttrpcutil.NewClient(containerdSockPath + ".ttrpc")
+	require.NoError(t, err, "failed to create ttrpc client")
+
+	image, err := alpineImage(ctx, client, naiveSnapshotterName)
+	require.NoError(t, err, "failed to get alpine image")
+
+	numVMs := 5
+	var vmIDs []string
+	webpages := make(map[string]string)
+	for i := 0; i < numVMs; i++ {
+		vmID := fmt.Sprintf("vm-%d", i)
+		vmIDs = append(vmIDs, vmID)
+		webpages[vmID] = fmt.Sprintf("Hello, my virtual machine %s\n", vmID)
+	}
+
+	localServices, err := internal.NewLocalNetworkServices(webpages)
+	require.NoError(t, err, "failed to create local network test services")
+
+	cniNetworkName := "fcnet-test"
+	err = writeCNIConf("/etc/cni/conf.d/fcnet-test.conflist", cniNetworkName, localServices.DNSServerIP())
+	require.NoError(t, err, "failed to write test cni conf")
+
+	go func() {
+		require.NoError(t, localServices.Serve(ctx), "failed serving local dns and http")
```

Comment:
This `require.NoError` might not actually catch an error that occurs later.  You might want to use a waitgroup and cancellable context here too.

```go
serveCtx, serveCancel := context.WithCancel(ctx)
var serveWait sync.WaitGroup
serveWait.Add(1)
go func() {
	defer serveWait.Done()
	require.NoError(t, localServices.Serve(ctx), "failed serving local dns and http")
}()

// ...

serveCancel()
serveWait.Wait()
```